packages/middleware-endpoint/src/endpointMiddleware.ts

```diff
@@ -1,24 +1,54 @@
+import { HttpRequest } from "@aws-sdk/protocol-http";
+import { parseQueryString } from "@aws-sdk/querystring-parser";
 import {
+  EndpointV2,
+  HandlerExecutionContext,
   MetadataBearer,
   SerializeHandler,
   SerializeHandlerArguments,
   SerializeHandlerOutput,
   SerializeMiddleware,
 } from "@aws-sdk/types";
 
-import { EndpointParameterInstruction } from "./types";
+import { getEndpointFromInstructions } from "./adaptors/getEndpointFromConfig";
+import { EndpointResolvedConfig } from "./resolveEndpointConfig";
+import { EndpointParameterInstructions } from "./types";
 
-export const endpointMiddleware = (options: {
-  config: any; // TODO(endpointsV2): should be ResolvedEndpointConfig interface
-  instruction: EndpointParameterInstruction;
+/**
+ * @private
+ */
+export const endpointMiddleware = ({
+  config,
+  instructions,
+}: {
+  config: EndpointResolvedConfig;
+  instructions: EndpointParameterInstructions;
 }): SerializeMiddleware<any, any> => {
-  return <Output extends MetadataBearer>(next: SerializeHandler<any, Output>): SerializeHandler<any, Output> =>
+  return <Output extends MetadataBearer>(
+      next: SerializeHandler<any, Output>,
+      context: HandlerExecutionContext
+    ): SerializeHandler<any, Output> =>
     async (args: SerializeHandlerArguments<any>): Promise<SerializeHandlerOutput<Output>> => {
-      // TODO(endpointsV2): resolve async providers from client config and static values according to
-      // instruction from input to populate endpoint parameters
+      if (HttpRequest.isInstance(args.request)) {
+        const { request } = args;
+
+        const endpoint: EndpointV2 = await getEndpointFromInstructions(args.input, instructions, config, context);
+
+        context.endpointV2 = endpoint;
+        context.authSchemes = endpoint.properties?.authSchemes;
+
+        request.headers = Object.entries(endpoint.headers || {}).reduce((headers, [name, values]) => {
+          headers[name] = values.join(",");
+          return headers;
+        }, {} as Record<string, string>);
```

Comment:
Can we changed to destructuring as follows
```suggestion
        request.headers = Object.entries(endpoint.headers || {}).reduce(
          (headers, [name, values]) => ({
            ...headers,
            [name]: values.join(","),
          }),
          {} as Record<string, string>
        );
```